programs/stake/src/stake_state.rs

```diff
@@ -1030,49 +1069,30 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
 
     fn merge(
         &self,
-        source_stake: &KeyedAccount,
+        source_account: &KeyedAccount,
         clock: &Clock,
         stake_history: &StakeHistory,
         signers: &HashSet<Pubkey>,
     ) -> Result<(), InstructionError> {
-        if source_stake.owner()? != id() {
+        if source_account.owner()? != id() {
             return Err(InstructionError::IncorrectProgramId);
         }
 
```

Comment:
we must check uniqueness check of `self` and `source_stake` (Well, I tried to find this kind of attack previously for all stake instructions... now that I've finally spotted one before releasing.)...

```
$ ./target/release/solana --url http://localhost:8899 stakes Bi8tSxKeSeUSCFpeKKhAkmL2kX1QuEcez1o9mJU3UrUT
Stake Pubkey: 8SE9zKZ1T51qtLiS8rSKcE29AHGnP66y95CrRWW2r11m
Balance: 2235671.215321532 SOL
Rent Exempt Reserve: 22.860288 SOL
Delegated Stake: 2235625.494745532 SOL
Active Stake: 2235625.494745532 SOL 
Delegated Vote Account Address: Bi8tSxKeSeUSCFpeKKhAkmL2kX1QuEcez1o9mJU3UrUT
Stake Authority: 7DEkZ3z7cWA7DpfBZjR6iCEdc5zkfr8wZi2TNBTkbt6j
Withdraw Authority: 7DEkZ3z7cWA7DpfBZjR6iCEdc5zkfr8wZi2TNBTkbt6j

$ ./target/release/solana --url http://localhost:8899 merge-stake 8SE9zKZ1T51qtLiS8rSKcE29AHGnP66y95CrRWW2r11m 8SE9zKZ1T51qtLiS8rSKcE29AHGnP66y95CrRWW2r11m
Signature: 4t9WfV5k4Ts7eq88dhDnS7FxhTp54wY6CK5LuiNKUZQAUd2YfmWwC8t7FCg8BCMidbqvKXoNPhhfYjtdc1b4HH2W

$ ./target/release/solana --url http://localhost:8899 stakes Bi8tSxKeSeUSCFpeKKhAkmL2kX1QuEcez1o9mJU3UrUT
Stake Pubkey: 8SE9zKZ1T51qtLiS8rSKcE29AHGnP66y95CrRWW2r11m
Balance: 2235676.240894693 SOL
Rent Exempt Reserve: 22.860288 SOL  
Delegated Stake: 4471261.040637386 SOL
Active Stake: 4471261.040637386 SOL 
Delegated Vote Account Address: Bi8tSxKeSeUSCFpeKKhAkmL2kX1QuEcez1o9mJU3UrUT
Stake Authority: 7DEkZ3z7cWA7DpfBZjR6iCEdc5zkfr8wZi2TNBTkbt6j
Withdraw Authority: 7DEkZ3z7cWA7DpfBZjR6iCEdc5zkfr8wZi2TNBTkbt6j

```

Reply:
Ha!  Nice!  I can add that check